nixos/public-inbox: fix test by creating Gitolite repositories

ju1m opened the pull request:
Previously in https://github.com/NixOS/nixpkgs/pull/308740 `BindReadOnlyPaths=` was fixed, but remained mounting non-existing Git repositories:

> vm-test-run-public-inbox> machine # [   19.503051] (ox-httpd)[1489]:
> public-inbox-httpd.service: Failed to set up mount namespacing:
> /var/lib/gitolite/repositories/user/repo1.git: No such file or directory


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

- [X] Create Git repositories (**without** using gitolite anymore).
- [X] Fix race condition between `curl` and `public-inbox-httpd.socket`

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

Ping @alyssais 

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ju1m committed:
nixos/public-inbox: fix test by creating Git repositories

Previously in https://github.com/NixOS/nixpkgs/pull/308740
`BindReadOnlyPaths=` was fixed, but remained mounting non-existing Git repositories:

vm-test-run-public-inbox> machine # [   19.503051] (ox-httpd)[1489]:
public-inbox-httpd.service: Failed to set up mount namespacing:
/var/lib/gitolite/repositories/user/repo1.git: No such file or directory

ju1m commented:
Hmm, why is OfBorg skipping the test ? https://github.com/NixOS/nixpkgs/pull/373522/checks?check_run_id=35548340273
```
OfBorg / nixos/public-inbox, nixos/public-inbox.passthru.tests on x86_64-linux skipped Jan 13, 2025 in 0s
No attempt

The following builds were skipped because they don't evaluate on x86_64-linux: nixos/public-inbox, nixos/public-inbox.passthru.tests

Details
Partial log

Cannot nix-instantiate `nixos/public-inbox` because:
error: attribute 'nixos/public-inbox' in selection path 'nixos/public-inbox' not found

Cannot nix-instantiate `nixos/public-inbox.passthru.tests` because:
error: attribute 'nixos/public-inbox' in selection path 'nixos/public-inbox.passthru.tests' not found
```

alyssais commented:
OfBorg parses attribute paths from commit messages, and "nixos/public-inbox" isn't one of them. (nixosTests.public-inbox would be though.)

@OfBorg test public-inbox

alyssais reviewed (approved):
I do wonder if it would be simpler to just use bare git repos that we could `git init` than using gitolite at all in this test, but if this makes it work again it's fine.

ju1m commented:
> OfBorg parses attribute paths from commit messages, and "nixos/public-inbox" isn't one of them. (nixosTests.public-inbox would be though.)

@alyssais Should I rather use `nixosTests.public-inbox` as a prefix of the commit message then?

ju1m commented:
https://github.com/ofborg test public-inbox

alyssais commented:
> > OfBorg parses attribute paths from commit messages, and "nixos/public-inbox" isn't one of them. (nixosTests.public-inbox would be though.)
> 
> @alyssais Should I rather use `nixosTests.public-inbox` as a prefix of the commit message then?

Up to you I think — I don't think it's much of a convention to do that for tests, but I haven't checked.

alyssais commented:
Any idea what's up with the failure on aarch64-linux?

ju1m committed:
nixos/public-inbox: fix flaky test by waiting for sockets

ju1m commented:
> Any idea what's up with the failure on aarch64-linux?

Looks like a race condition between the `curl` and the `public-inbox-httpd` upstream:

> nginx[1097]: 2025/01/16 19:26:04 [error] 1097#1097: *1 upstream timed out (110: Connection timed out) while reading response header from upstream, client: 2001:db8:1::1, server: machine.example.localdomain, request: "GET /inbox/ HTTP/2.0", upstream: "http://unix:/run/public-inbox-http.sock/inbox/", host: "machine.example.localdomain"

I've added a wait for the socket before the `curl` to `nginx`:
```python
machine.wait_for_unit("public-inbox-httpd.socket")
```
And removed the useless:
```python
machine.wait_for_unit("public-inbox-httpd.service")
```

@ofborg test public-inbox

alyssais commented:
@OfBorg test public-inbox

alyssais commented:
Still failing…

ju1m commented:
> Still failing…

Hm, no more "upstream timed out", just a curl timeout AFAICS, more verbose logs for curl and journal of public-inbox-httpd would help. So next step would be to reproduce locally using qemu-system-aarch64. Not sure I'll dig that hole though.

alyssais commented:
It works for me on aarch64 locally, so I don't think it's architecture specific.  It must be a timing thing.
